### PR TITLE
get thresholding result from ercf

### PR DESCRIPTION
**Pull request by wnbts**
### Description
This pr uses ercf to get anomaly result.
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/ml/ModelManager.java [hunk 1]
```diff
@@ -817,20 +819,25 @@ public ThresholdingResult getAnomalyResultForEntity(
...
 
-            // trainModelFromExistingSamples may be able to make models not null
-            if (entityModel.getRcf() == null || entityModel.getThreshold() == null) {
-                entityColdStarter.trainModelFromExistingSamples(modelState);
-            }
-
-            if (entityModel.getRcf() != null && entityModel.getThreshold() != null) {
-                return score(datapoint, modelId, modelState);
+            if (entityModel.getErcf().isPresent()) {
+                result = toResult(entityModel.getErcf().get().process(datapoint));
```
Comment (cswiercz) [inline src/main/java/org/opensearch/ad/ml/ModelManager.java:823]: Design nitpick: the original code (below) uses early returns. The new code here is saying that if an ERCF exists then we will prefer it's output as opposed to the "original model" output. Why not do an early return here? It's allowed since this is the only line that sets `result`. It will decrease the cognitive load, a bit.

Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:823]: * yes, ercf will be the current/only model in the system.
* multiple returns are considered less clear than a single return so the new code switches to that. and old code for old models will be removed because with later changes, old models will be converted to ercf before use.

Comment (cswiercz) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:823]: Sounds good.

**Review by cswiercz — APPROVED** [review]

File: src/test/java/org/opensearch/ad/ml/ModelManagerTests.java [hunk 2]
```diff
@@ -266,6 +277,9 @@ public void setup() {
         detectorId = "detectorId";
         rcfModelId = "detectorId_model_rcf_1";
         thresholdModelId = "detectorId_model_threshold";
+
+        when(this.modelState.getModel()).thenReturn(this.entityModel);
+        when(this.entityModel.getErcf()).thenReturn(Optional.of(this.ercf));
```
Comment (kaituo) [inline src/test/java/org/opensearch/ad/ml/ModelManagerTests.java:282]: This would make it hard to test the else branch of entityModel.getErcf().isPresent() == false.  Can we move it to somewhere else?

File: src/main/java/org/opensearch/ad/ml/ModelManager.java [hunk 3]
```diff
@@ -817,20 +819,25 @@ public ThresholdingResult getAnomalyResultForEntity(
...
+                // trainModelFromExistingSamples may be able to make models not null
+                if (entityModel.getRcf() == null || entityModel.getThreshold() == null) {
+                    entityColdStarter.trainModelFromExistingSamples(modelState);
+                }
+
+                if (entityModel.getRcf() != null && entityModel.getThreshold() != null) {
+                    return score(datapoint, modelId, modelState);
+                } else {
+                    entityModel.addSample(datapoint);
+                    return new ThresholdingResult(0, 0, 0);
```
Comment (kaituo) [inline src/main/java/org/opensearch/ad/ml/ModelManager.java:834]: Seems line 834 and 838 can be removed as those are the last line of the branches and they are doing the same thing of line 840.

Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java]: yes, they can be removed.

File: src/test/java/org/opensearch/ad/ml/ModelManagerTests.java (hunk 2 above)
Comment (wnbts) [reply src/test/java/org/opensearch/ad/ml/ModelManagerTests.java:282]: it shouldn't not affect new test cases since new test cases can always override this behavior so this is just a default for convenience.

File: src/main/java/org/opensearch/ad/ml/ModelManager.java [hunk 4]
```diff
@@ -817,20 +819,22 @@ public ThresholdingResult getAnomalyResultForEntity(
...
+                // trainModelFromExistingSamples may be able to make models not null
+                if (entityModel.getRcf() == null || entityModel.getThreshold() == null) {
+                    entityColdStarter.trainModelFromExistingSamples(modelState);
+                }
+
+                if (entityModel.getRcf() != null && entityModel.getThreshold() != null) {
+                    return score(datapoint, modelId, modelState);
+                } else {
+                    entityModel.addSample(datapoint);
+                }
```
Comment (LiuJoyceC) [inline src/main/java/org/opensearch/ad/ml/ModelManager.java:834]: [Clarification question] Just to confirm, does invoking `entityColdStarter.trainModelFromExistingSamples(modelState)` synchronously change the result of `entityModel.getRcf()` and/or `entityModel.getThreshold()`? Otherwise the second if statement is redundant.

[Minor suggestion, non-blocking] Since this set of states is being checked twice (just in opposite cases), it would be more elegant and easier to read to add a method on `entityModel` that checks this exact condition. Then you could have:
```
if (entityModel.hasRcfOrThreshold()) { ... }
if (!entityModel.hasRcfOrThreshold() { ... } else { ... }
```

File: src/test/java/org/opensearch/ad/ml/ModelManagerTests.java [hunk 5]
```diff
@@ -973,4 +987,19 @@ public void getPreviewResults_returnAnomalies_forLastAnomaly() {
...
+        anomalyDescriptor.setAnomalyGrade(1);
+        when(this.ercf.process(this.point)).thenReturn(anomalyDescriptor);
+
+        ThresholdingResult result = modelManager
+            .getAnomalyResultForEntity(this.point, this.modelState, this.detectorId, this.anomalyDetector, null);
+        assertEquals(
+            new ThresholdingResult(anomalyDescriptor.getAnomalyGrade(), /*TODO: pending ercf*/1.0, anomalyDescriptor.getRcfScore()),
+            result
+        );
+    }
```
Comment (LiuJoyceC) [inline src/test/java/org/opensearch/ad/ml/ModelManagerTests.java:1004]: Instead of hardcoding the value 2 for the RCF score and 1 for the anomaly grade, this test function could accept arguments for the RCF score and anomaly grade. Then you can test a few different combinations of values, including edge cases where an unexpected behavior could happen.

Comment (LiuJoyceC) [reply src/test/java/org/opensearch/ad/ml/ModelManagerTests.java:1004]: What is `/*TODO: pending ercf*/`? Is that waiting on another PR?

File: src/main/java/org/opensearch/ad/ml/ModelManager.java (hunk 4 above)
Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:834]: 1. Both should be set or unset. But it's fine and the code for old models will be removed soon as it is replaced by new ercf.
2. As above, this branch will be left as it is and soon be replaced by new logic based on ercf.

File: src/test/java/org/opensearch/ad/ml/ModelManagerTests.java (hunk 5 above)
Comment (wnbts) [reply src/test/java/org/opensearch/ad/ml/ModelManagerTests.java:1004]: 1. Parameterized testing is useful. In this case, the logic is quite simple as just copying data so the test can therefore be simple. If the logic becomes more complex, the corresponding tests can be parameterized to scale to the test cases. However, that is not always needed.
2. The comment is a TODO item for the confidence part of ercf is still in development and not available yet. It unblocks development and integration on the plugin side for the time being.

**Review by kaituo — APPROVED** [review]

File: src/main/java/org/opensearch/ad/ml/ModelManager.java (hunk 1 above)
Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:823]: 1. Can the `process` method support batch data? Asking this as we have historical analysis which run in batch. 
2. Do you have any suggestion about how many data points should be good enough for ERCF to generate results with high confidence.

File: src/main/java/org/opensearch/ad/ml/ModelManager.java [hunk 6]
```diff
@@ -817,20 +819,22 @@ public ThresholdingResult getAnomalyResultForEntity(
...
-
-            if (entityModel.getRcf() != null && entityModel.getThreshold() != null) {
-                return score(datapoint, modelId, modelState);
+            if (entityModel.getErcf().isPresent()) {
+                result = toResult(entityModel.getErcf().get().process(datapoint));
             } else {
-                entityModel.addSample(datapoint);
-                return new ThresholdingResult(0, 0, 0);
+                // trainModelFromExistingSamples may be able to make models not null
+                if (entityModel.getRcf() == null || entityModel.getThreshold() == null) {
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/ml/ModelManager.java:826]: Should we remove such RCF and Threshold things as ERCF hide these behind the scene ?

File: src/main/java/org/opensearch/ad/ml/ModelManager.java (hunk 1 above)
Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:823]: 1. No, it only takes one point at a time. For a batch of data points, the client code calls on each of them.
2. Not new numbers yet. The current number should be sufficient.

File: src/main/java/org/opensearch/ad/ml/ModelManager.java (hunk 6 above)
Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/ModelManager.java:826]: Yes, eventually. This branch dealing with existing models will be replaced with new pr.

**Review by LiuJoyceC — APPROVED** [review]
